bb8/src/inner.rs

```diff
@@ -194,6 +209,22 @@ where
             }
         }
     }
+
+    async fn on_acquire_connection(
+        &self,
+        mut conn: M::Connection,
```

Comment:
Should this take a `&mut` as well, rather than taking ownership? The semantics here can be tricky since a future might get dropped during an await point, in which case we'd leak the connection here.

Reply:
That was my first choice too but I couldn't get it to compile that way (and the semantics nuance is a bit lost on me). 

Here is what it looks like:
https://github.com/agersant/bb8/commit/2fd5045e5a49c589047409df4de92bf8b8701c86
(or [compacted](https://github.com/agersant/bb8/commit/50945f6f5a1fee914cd0fb2c50852b577548fce1) with `and_then`)

The borrow checker is not happy with either of these and I'm not sure how to fix that.

![image](https://user-images.githubusercontent.com/817256/101999545-09f76780-3c93-11eb-8b79-bed943fb5db4.png)

Reply:
Can you try writing it like this?

```rust
let conn = self.inner.manager.connect().await;
self.on_acquire_connection(&mut conn).await?;
Ok(conn)
```

Reply:
Haha, that works fine. Less is more!

The other call site looks a little less nice, but I got it to compile but just manually unrolling the `map` logic.